Add a flag to infer appropriate cross compile hosts on Darwin

edymtt opened the pull request:
Use that immediately for the macOS package presets.

Addresses rdar://98787335, #60348

edymtt commented:
This is based off the suggestion from @shahmishal in #60624

Keeping this as a draft since I expect to iterate on tests and linting.

edymtt commented:
@swift-ci please python lint

compnerd reviewed (approved):
I'm not sure about the syntax for the macOS test, but if the CI is happy with it, that seems reasonable.  This also addresses my question in the original PR.

edymtt commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -580,6 +580,11 @@ def create_argument_parser():
+    option('--infer-cross-compile-hosts-on-darwin', toggle_true,
+           help="When building on Darwin, populates automatically cross-compile-hosts "
+                "based on the architecture build-script is running on. "
+                "Has precendence over cross-compile-hosts")
```
```suggestion
                "Has precedence over cross-compile-hosts")
```

edymtt replied:
Tackled in c954130

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

edymtt commented:
@swift-ci please build Apple Silicon

edymtt committed:
Add a flag to infer appropriate cross compile hosts on Darwin

Addresses rdar://98787335, apple#60348

edymtt committed:
Infer automatically cross compile hosts for macOS package presets

Addresses rdar://98787335, apple#60348

edymtt commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -580,6 +580,11 @@ def create_argument_parser():
                 'Swift tools for. Can be used multiple times.')
 
+    option('--infer-cross-compile-hosts-on-darwin', toggle_true,
+           help="When building on Darwin, populates automatically cross-compile-hosts "
```
Thanks -- tackled in c954130

shahmishal reviewed (approved):
Thanks for changing this to dynamic.

edymtt commented:
* fixed `validation-test/BuildSystem/infer-cross-compile-hosts-on-darwin-macosx.test` to rely instead on `%target-cpu` -- the `%if` feature I was using is not present in `stable/20220421` (https://reviews.llvm.org/rG1041a9642ba035fd2685f925911d705e8edf5bb0)
* split the usage of the new flag in a separate commit, so in case of regressions that can be reverted separately while leaving the bulk of the logic in place (but inert)

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

edymtt commented:
@swift-ci please build Apple Silicon

edymtt commented:
@swift-ci please python lint

drexin reviewed (approved):
LGTM, thanks for doing this 🙂

edymtt commented:
Hitting an unrelated failure in macOS PR testing now solved by #60709, retrying

@swift-ci please macOS

rayxia commented:
@edymtt Hi, after using this code, I got another error when buidling the toolchain, can you help me?

```
Undefined symbols for architecture arm64:
  "llvm::cl::Option::anchor()", referenced from:
      vtable for llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > in swift-demangle.cpp.o
  "llvm::cl::Option::addOccurrence(unsigned int, llvm::StringRef, llvm::StringRef, bool)", referenced from:
      vtable for llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > in swift-demangle.cpp.o
  "llvm::cl::GenericOptionValue::anchor()", referenced from:
      vtable for llvm::cl::OptionValueCopy<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > in swift-demangle.cpp.o
      vtable for llvm::cl::OptionValue<bool> in swift-demangle.cpp.o
  "vtable for llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::Option::addArgument()", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  "vtable for llvm::cl::OptionValue<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::Option::setArgStr(llvm::StringRef)", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  "llvm::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long)", referenced from:
      llvm::cl::Option::Option(llvm::cl::NumOccurrencesFlag, llvm::cl::OptionHidden) in swift-demangle.cpp.o
  "vtable for llvm::cl::opt<bool, false, llvm::cl::parser<bool> >", referenced from:
      llvm::cl::opt<bool, false, llvm::cl::parser<bool> >::~opt() in swift-demangle.cpp.o
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::basic_parser_impl::getOptionWidth(llvm::cl::Option const&) const", referenced from:
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::getOptionWidth() const in swift-demangle.cpp.o
  "llvm::cl::ParseCommandLineOptions(int, char const* const*, llvm::StringRef, llvm::raw_ostream*, char const*, bool)", referenced from:
      _main in swift-demangle.cpp.o
  "llvm::raw_ostream::write(unsigned char)", referenced from:
      _main in swift-demangle.cpp.o
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(char) in swift-demangle.cpp.o
  "vtable for llvm::cl::parser<bool>", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::getGeneralCategory()", referenced from:
      llvm::cl::Option::Option(llvm::cl::NumOccurrencesFlag, llvm::cl::OptionHidden) in swift-demangle.cpp.o
  "llvm::cl::basic_parser_impl::printOptionInfo(llvm::cl::Option const&, unsigned long) const", referenced from:
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::printOptionInfo(unsigned long) const in swift-demangle.cpp.o
  "llvm::outs()", referenced from:
      _main in swift-demangle.cpp.o
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
  "llvm::errs()", referenced from:
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
  "llvm::raw_ostream::write(char const*, unsigned long)", referenced from:
      _main in swift-demangle.cpp.o
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(char const*) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in swift-demangle.cpp.o
  "vtable for llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >", referenced from:
      llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~opt() in swift-demangle.cpp.o
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::raw_ostream::operator<<(long)", referenced from:
      llvm::raw_ostream::operator<<(int) in swift-demangle.cpp.o
  "llvm::raw_ostream::operator<<(unsigned long)", referenced from:
      llvm::raw_ostream::operator<<(unsigned int) in swift-demangle.cpp.o
  "vtable for llvm::cl::Option", referenced from:
      llvm::cl::opt<bool, false, llvm::cl::parser<bool> >::~opt() in swift-demangle.cpp.o
      llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~opt() in swift-demangle.cpp.o
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~list() in swift-demangle.cpp.o
      llvm::cl::Option::Option(llvm::cl::NumOccurrencesFlag, llvm::cl::OptionHidden) in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture arm64
```

edymtt commented:
@rayxia this is not something I hit so far when building on Apple Silicon

I would suggest creating an issue at https://github.com/apple/swift/issues/new/choose
providing at least
* the machine you are building on (OS and architecture)
* the build-script/build-toolchain invocation you are using

If possible it would be nice to have
* the full log from the run
* the build.ninja files from any build/*/swift-macosx-<arch> folders you have

(not sure how much I can follow up since I'm busy with other efforts, but this way you others may be more likely to see your issue)

rayxia commented:
> @rayxia this is not something I hit so far when building on Apple Silicon
> 
> I would suggest creating an issue at https://github.com/apple/swift/issues/new/choose providing at least
> 
> * the machine you are building on (OS and architecture)
> * the build-script/build-toolchain invocation you are using
> 
> If possible it would be nice to have
> 
> * the full log from the run
> * the build.ninja files from any build/*/swift-macosx- folders you have
> 
> (not sure how much I can follow up since I'm busy with other efforts, but this way you others may be more likely to see your issue)

@edymtt OK, thanks for your suggestion.
